### PR TITLE
[docs] Generate all :board docs explicitly

### PR DESCRIPTION
**Pull request by salkinium**
I discovered that we only generate the docs for the largest memory devices, thus missing most of the targets that contain the board modules. I'm adding them here explicitly, both in the module docs, as well as in the doxypress docs.

I'm merging the devices and boards like this: `docs.modm.io/develop/api/blue-pill` and `docs.modm.io/develop/api/stm32f103c8t6`, is this acceptable?

I also removed the `Version/Release` input, since we're not doing that yet.

- [x] Verify module docs has all board modules
- [x] Verify doxypress docs has all board modules
- [x] Fix broken dependency graphs in module docs
- [x] Simple deduplication via symlink

cc @rleh

### CODE REVIEW

File: tools/scripts/docs_modm_io_index.html.in [hunk 1]
```diff
@@ -60,7 +60,6 @@ input[type=submit] {
     top: 100%;
     left: 0;
     right: 0;
-    text-transform: uppercase;
```
Comment (salkinium) [inline tools/scripts/docs_modm_io_index.html.in:63]: I spend like 2mins trying to undo CAPLOCK until I realized that this was CSS ;-P

Comment (rleh) [reply tools/scripts/docs_modm_io_index.html.in:63]: IMHO this is necessary for the case-insensitive search to function correctly.

**Review by rleh — CHANGES_REQUESTED** [review]

File: tools/scripts/docs_modm_io_index.html.in [hunk 2]
```diff
@@ -85,24 +84,26 @@ input[type=submit] {
...
         <td>
             <select id="releaseinput" name="releases">
                 <option>develop</option>
             </select>
         </td>
+        -->
         <td>
             <div class="autocomplete" style="width:300px;">
-                <input id="targetinput" type="text" name="target" placeholder="Type e.g. 'F407'">
+                <input id="targetinput" type="text" name="target" placeholder="Type e.g. 'F407' or 'Blue Pill'">
```
Comment (rleh) [inline tools/scripts/docs_modm_io_index.html.in]:
```suggestion
                <input id="targetinput" type="text" name="target" placeholder="Type e.g. 'F407' or 'Nucleo-G474'">
```

Not everyone immediately recognizes "Blue Pill" as a dev board....

File: tools/scripts/docs_modm_io_index.html.in [hunk 3]
```diff
@@ -219,8 +232,11 @@ document.addEventListener("click", function (element) {
 </script>
 <noscript>
 <h3>Select your target</h3>
-<input type="text" id="input" placeholder="Type e.g. 'F407'" />
+<input type="text" id="input" placeholder="Type e.g. 'F407' or 'Blue Pill'"/>
```
Comment (rleh) [inline tools/scripts/docs_modm_io_index.html.in]: Same as above

File: tools/scripts/docs_modm_io_index.html.in (hunk 2 above)
Comment (salkinium) [reply tools/scripts/docs_modm_io_index.html.in]: Perhaps just `Search for device or board` and the people just know it's a search field?

Comment (rleh) [reply tools/scripts/docs_modm_io_index.html.in]: Nah, having an example is always better.

Comment (salkinium) [reply tools/scripts/docs_modm_io_index.html.in]: Yeah, it also just duplicates the description, which already says `Device or board`.

File: tools/scripts/docs_modm_io_index.html.in (hunk 1 above)
Comment (salkinium) [reply tools/scripts/docs_modm_io_index.html.in:63]: No, you already thought about this and did `.toLowerCase()` on both the input and the device list in the function. I left the device upper case though, just made the boards mixed cases as their official names.

**Comment by salkinium** [comment]
I also noticed the module dependency graph is not shown on the homepage

File: tools/scripts/docs_modm_io_generator.py [hunk 4]
```diff
@@ -92,13 +104,15 @@ def main():
     args = parser.parse_args()
 
     device_list = []
+    board_list = []
     if args.test:
-        # test list
         device_list = ["hosted-linux", "atmega328p-au", "stm32f103c8t6", "stm32g474cet6", "samd21g18a-uu"]
+        board_list = [("arduino-nano", "atmega328p-au"), ("arduino-uno", "atmega328p-au"), ("nucleo-g474re", "stm32g474ret6"),
+                      ("blue-pill", "stm32f103c8t6"), ("feather-m0", "samd21g18a-uu")]
```
Comment (rleh) [inline tools/scripts/docs_modm_io_generator.py:112]: 👍🏽

**Review by rleh — APPROVED** [review]

**Comment by rleh** [comment]
Do you know how much additional storage is required for the added targets?

**Comment by salkinium** [comment]
> Do you know how much additional storage is required for the added targets?

It's adding an entire doxypress folder per board, so 45 addition folders. I don't know exactly how much this adds, but when I ran this locally on my machine the total was 10GB uncompressed.

It's not a particularly efficient solution.

**Comment by rleh** [comment]
Currently we have approximately 13GB (uncompressed) from 141 targets.
The server has 40GB dedicated to https://docs.modm.io/, the uncompressed data and the gzip upload must fit in there together.

> It's not a particularly efficient solution.

Depends on the definition of "efficiency". The cost of (cold) storage is very low, I am not aware of a better solution at the moment...

**Comment by rleh** [comment]
Merge or do you want to fix the module dependency graph is not shown on the homepage?

**Comment by salkinium** [comment]
I want to fix the dependency graph first, I can already reproduce it locally. Something related to mkdocs upgrade.

I have a feeling we could do some deduplication for identical files afterwards via softlinks. I'll check that out too. Perhaps it's a simple solution.

**Comment by rleh** [comment]
> I have a feeling we could do some deduplication for identical files afterwards via softlinks. I'll check that out too. Perhaps it's a simple solution.

Webservers typically don't follow symlinks for security reasons.

**Comment by salkinium** [comment]
That's a shame because most files are actually byte-for-byte identical, with the exception of the title (trivially fixable).

For example the `modm::amnb::Device` is device-independent also in documentation:
```
 $ diff develop/api/arduino-uno/classmodm_1_1amnb_1_1Device.html develop/api/nucleo-g474re/classmodm_1_1amnb_1_1Device.html
8c8
< <title>modm API documentation for ATMEGA328P-AU: modm::amnb::Device Class Reference</title>
---
> <title>modm API documentation for STM32G474RET6: modm::amnb::Device Class Reference</title>
31c31
<    <div id="projectname">modm API documentation for ATMEGA328P-AU
---
>    <div id="projectname">modm API documentation for STM32G474RET6
```

But the `modm::Adc` class is different:

```
 $ diff develop/api/arduino-uno/classmodm_1_1Adc.html develop/api/nucleo-g474re/classmodm_1_1Adc.html
8c8
< <title>modm API documentation for ATMEGA328P-AU: modm::Adc Class Reference</title>
---
> <title>modm API documentation for STM32G474RET6: modm::Adc Class Reference</title>
31c31
<    <div id="projectname">modm API documentation for ATMEGA328P-AU
---
>    <div id="projectname">modm API documentation for STM32G474RET6
117c117
< <p><code>#include &lt;<a class="el" href="architecture_2interface_2adc_8hpp_source.html">modm/architecture/interface/adc.hpp</a>&gt;</code></p>
---
> <p><code>#include &lt;<a class="el" href="adc_8hpp_source.html">modm/architecture/interface/adc.hpp</a>&gt;</code></p>
124,128c124,125
< <area href="classmodm_1_1PeripheralDriver.html" alt="modm::PeripheralDriver" shape="rect" coords="90,0,261,24"/>
< <area href="classmodm_1_1AdcInterrupt.html" alt="modm::AdcInterrupt" shape="rect" coords="0,112,171,136"/>
< <area href="classmodm_1_1platform_1_1Adc.html" alt="modm::platform::Adc" shape="rect" coords="181,112,352,136"/>
< <area href="classmodm_1_1platform_1_1AdcInterrupt.html" alt="modm::platform::AdcInterrupt" shape="rect" coords="0,168,171,192"/>
< <area href="classmodm_1_1platform_1_1AdcInterrupt.html" alt="modm::platform::AdcInterrupt" shape="rect" coords="181,168,352,192"/>
---
> <area href="classmodm_1_1PeripheralDriver.html" alt="modm::PeripheralDriver" shape="rect" coords="0,0,145,24"/>
> <area href="classmodm_1_1AdcInterrupt.html" alt="modm::AdcInterrupt" shape="rect" coords="0,112,145,136"/>
462c459
< <li><a class="el" href="architecture_2interface_2adc_8hpp_source.html">architecture/interface/adc.hpp</a></li>
---
> <li><a class="el" href="adc_8hpp_source.html">adc.hpp</a></li>
```

**Comment by rleh** [comment]
We can give it a try, I can take a look at the [webserver config](http://nginx.org/en/docs/http/ngx_http_core_module.html#disable_symlinks) and make changes if necessary.

---

1. Remove target identifier from Doxypress projectname. (Later add the target name to HTML again via Javascript, e.g. extract it from the URI...)
2. Compute hashes of all files and then replace all identical files?

Step 2 should be optional (command line flag to generator script), because maybe in the future we will use [S3](https://en.wikipedia.org/wiki/Amazon_S3)-like storage services that don't support symlinks (and maybe have transparent duplication like some modern filesystems)

**Comment by salkinium** [comment]
I'll hack something together, I think it can be done on the fly, so the disk space requirement for generating the docs should be low too. I first have to fix the damn inline SVG issue…

**Comment by salkinium** [comment]
Ok, the solution is fairly simple, but it's now all synchronously run, since doing this over multiprocessing is a bit difficult with a shared dictionary.

If you want to change your server config, I can push to a `develop-doc-dedup` branch and it'll upload to your server.

**Comment by salkinium** [comment]
Running this locally I get a ~3GB total folder (previously 10GB) and ~450MB tarfile.

**Comment by rleh** [comment]
> Later add the target name to HTML again via Javascript, e.g. extract it from the URI...

Done in de2d3dd.

**Review by rleh — APPROVED** [review]

**Comment by rleh** [comment]
> If you want to change your server config, I can push to a `develop-doc-dedup` branch and it'll upload to your server.

The webserver should now follow symlinks. I did a push to `develop-doc-dedup` so let's wait...

**Comment by salkinium** [comment]
> I did a push to develop-doc-dedup so let's wait...

Well, the homepage module docs are already fixed. All of the boards are there and the dependency graph is too.

**Comment by salkinium** [comment]
An error unrelated to these changes:
```
Error generating documentation for device raspberrypi: In 'modm-test:test':

Traceback (most recent call last):
  File "/root/project/test/module.lb", line 26, in build
    core = next(t for t in ("avr", "cortex-m", "hosted") if core.startswith(t))
StopIteration
```

**Comment by salkinium** [comment]
It's looks like the symlinks are working fine, however, the URLs are case sensitive:
https://doc.modm.io/develop/api/stm32f030rct6/ vs. https://doc.modm.io/develop/api/STM32F030RCT6/

**Comment by salkinium** [comment]
(I'm fixing it)

File: tools/scripts/docs_modm_io_index.html.in [hunk 5]
```diff
@@ -143,6 +152,9 @@ function showDocumentation() {
         return;
     }
     var url = "/" + releaseinput.value + "/api/" + targetinput.value + "/";
+    */
+    n2b = name2board[targetinput.value]
+    var url = "/develop/api/" + (n2b ? n2b : targetinput.value) + "/";
```
Comment (rleh) [inline tools/scripts/docs_modm_io_index.html.in]: This should fix it?!

```suggestion
    var url = "/develop/api/" + (n2b ? n2b : targetinput.value).toLowerCase() + "/";
```

Comment (salkinium) [reply tools/scripts/docs_modm_io_index.html.in]: Yes, that's what I added.

**Comment by salkinium** [comment]
Pushed to develop-doc-dedup again, let's see this again in 1h15.

**Comment by salkinium** [comment]
It's working, incl. RPi docs! \o/
What's the space saving on the server?

**Comment by rleh** [comment]
>  What's the space saving on the server?

Currently only 3.3GB are used ... and a 4.8GB log file (which was also yesterday already there) 🤣

**Review by rleh — APPROVED** [review]
Ready to merge!

**Pull request merged**